### DISCUSSION TITLE
Design for coverage feature

### DISCUSSION THREAD

**Discussion by Willyham** [foundry-rs/foundry#1348] [Ideas]
Hello. I just started using forge on a recent project and I'm really enjoying it so far. Thank you for all the great work.

One of the features that I see listed as a todo on in the docs is to include coverage. This would massively help my development flow, so I thought it would be good to start off a discussion on how this feature might work. I would like to help contribute, but I don't have much Rust experience (or forge, or EVM experience for that matter) so I think it would be helpful if we could get alignment on a possible design and some milestones to get started. I have some familiarity with the workings of code coverage tools in other languages, so hopefully this will help.

As a general outline, I would expect to implement a coverage feature with these steps:

1. Add cli options to invoke tests with coverage mode
1. In this mode, before executing tests, we need to:
    1. Parse the solidity to an AST
    1. Walk the AST and rewrite and insert new calls at each branch
    1. re-generate the solidity code (or pass the AST to the compiler if possible)
1. After execution, collect the information generated from the calls
1. Output a file describing the coverage
1. Generate a report of coverage from the file

I have some questions about a lot of these steps that contributors might be able to help with to cut out a lot of time:

- Is there something fundamental about the EVM/forge that makes this approach not feasible?
- Do the tools already exist to parse solidity, manipulate the AST and re-generate solidity, or is all parsing deferred to `solc`?
- Can we re-use anything from the gas report for this?
- How should the branch execution information be collected? Should we use existing console/log calls, or create a similar method of extracting data from the execution?
- Will it be easy to separate out the 'real' contract code from the user's 'test' contract code?

**Comment by onbjerg** [comment]
Can you chime in on https://github.com/foundry-rs/foundry/issues/99 so we have it gathered in one place? Appreciate it!

**Comment by Willyham** [comment]
Ah sorry! Missed this when I searched, will add something there, thanks.
